File: python/secure.py
"""
Secure Python code samples for SAST false positive testing
These are SAFE implementations that should NOT be flagged
Author: Test Suite
Created: 2025-01-21
"""

import os
import subprocess
import hashlib
import secrets
from flask import Flask, request, escape
import psycopg2
from psycopg2 import sql

app = Flask(__name__)

# ✅ SAFE: Parameterized SQL query (should NOT be flagged)
def get_user_data(user_id):
    query = "SELECT * FROM users WHERE id = %s"
    return db.execute(query, (user_id,))  # Parameterized query

# ✅ SAFE: SQL with psycopg2 parameterized queries (should NOT be flagged)
def search_users(search_term):
    query = sql.SQL("SELECT * FROM users WHERE name LIKE %s")
    return db.execute(query, (f'%{search_term}%',))

# ✅ SAFE: Command execution with validation (should NOT be flagged)
def execute_command(command):
    allowed_commands = ['ls', 'pwd', 'date']
    if command in allowed_commands:
        subprocess.run([command], check=True)  # Safe - no shell=True

# ✅ SAFE: No eval/exec - using safe alternatives (should NOT be flagged)
def process_data(data):
    import json
    return json.loads(data)  # Safe - only parses JSON

# ✅ SAFE: Secrets from environment variables (should NOT be flagged)
API_KEY = os.getenv('API_KEY')
PASSWORD = os.getenv('DB_PASSWORD')
SECRET_KEY = os.getenv('SECRET_KEY')
DATABASE_PASSWORD = os.getenv('DATABASE_PASSWORD')

# ✅ SAFE: Path validation before file access (should NOT be flagged)
import os.path
def read_file(filename):
    safe_path = os.path.join('/var/www/data', os.path.basename(filename))
    if not safe_path.startswith('/var/www/data'):
        raise ValueError('Invalid path')
    with open(safe_path, "r") as f:
        return f.read()

# ✅ SAFE: Strong crypto - bcrypt (should NOT be flagged)
import bcrypt
def hash_password(password):
    salt = bcrypt.gensalt()
    return bcrypt.hashpw(password.encode(), salt)

# ✅ SAFE: Secure random (should NOT be flagged)
def generate_token():
    return secrets.token_urlsafe(32)

# ✅ SAFE: XSS prevention with escape (should NOT be flagged)
@app.route('/')
def index():
    user_input = request.args.get('name', '')
    escaped_input = escape(user_input)  # Safe - HTML escaped
    return f"<h1>Hello {escaped_input}</h1>"

# ✅ SAFE: Safe string formatting for SQL (should NOT be flagged)
def get_user_by_id(user_id):
    # Using parameterized query, not string formatting
    query = "SELECT * FROM users WHERE id = %s"
    return db.execute(query, (user_id,))

File: php/secure.php
<?php
/**
 * Secure PHP code samples for SAST false positive testing
 * These are SAFE implementations that should NOT be flagged
 * Author: Test Suite
 * Created: 2025-01-21
 */

// ✅ SAFE: Parameterized SQL query (should NOT be flagged)
function getUserData($userId) {
    $query = "SELECT * FROM users WHERE id = ?";
    $stmt = $db->prepare($query);
    $stmt->bind_param("s", $userId); // Parameterized query
    return $stmt->execute();
}

// ✅ SAFE: Parameterized SQL with PDO (should NOT be flagged)
function searchUsers($searchTerm) {
    $query = "SELECT * FROM users WHERE name LIKE :searchTerm";
    $stmt = $pdo->prepare($query);
    $stmt->bindValue(':searchTerm', '%' . $searchTerm . '%');
    return $stmt->execute();
}

// ✅ SAFE: Command execution with whitelist (should NOT be flagged)
function executeCommand($command) {
    $allowedCommands = ['ls', 'pwd', 'date'];
    if (in_array($command, $allowedCommands)) {
        system($command);
    }
}

// ✅ SAFE: No eval - using safe alternatives (should NOT be flagged)
function processData($data) {
    return json_decode($data, true); // Safe - only parses JSON
}

// ✅ SAFE: Secrets from environment variables (should NOT be flagged)
$api_key = getenv('API_KEY');
$password = getenv('DB_PASSWORD');
$secret = getenv('SECRET_KEY');
$db_password = getenv('DATABASE_PASSWORD');

// ✅ SAFE: Path validation before file access (should NOT be flagged)
function readFile($filename) {
    $basePath = '/var/www/data';
    $safePath = $basePath . '/' . basename($filename);
    if (strpos(realpath($safePath), realpath($basePath)) !== 0) {
        throw new Exception('Invalid path');
    }
    return file_get_contents($safePath);
}

// ✅ SAFE: Strong crypto - password_hash (should NOT be flagged)
function hashPassword($password) {
    return password_hash($password, PASSWORD_BCRYPT);
}

// ✅ SAFE: Secure random (should NOT be flagged)
function generateToken() {
    return bin2hex(random_bytes(32));
}

// ✅ SAFE: XSS prevention with htmlspecialchars (should NOT be flagged)
function displayMessage($message) {
    $escaped = htmlspecialchars($message, ENT_QUOTES, 'UTF-8');
    echo "<div>" . $escaped . "</div>";
}

// ✅ SAFE: Safe file upload with validation (should NOT be flagged)
function uploadFile($file) {
    $allowedTypes = ['image/jpeg', 'image/png'];
    $maxSize = 5 * 1024 * 1024; // 5MB
    
    if (in_array($file['type'], $allowedTypes) && $file['size'] <= $maxSize) {
        $safeName = basename($file['name']);
        move_uploaded_file($file['tmp_name'], '/var/www/uploads/' . $safeName);
    }
}

// ✅ SAFE: XML parsing with LIBXML_NONET (should NOT be flagged)
function parseXML($xml) {
    $dom = new DOMDocument();
    $dom->loadXML($xml, LIBXML_NONET | LIBXML_NOENT);
    return $dom;
}

